build_tools/github_actions/runner/instance_deleter/main.py

```diff
@@ -110,27 +115,114 @@ def _get_from_items(items: compute.Items, key: str):
+        region=region,
+        # For some reason we can't just use a list of instance names and need to
+        # build this RhymingRythmicJavaClasses proto. Also, unlike all the other
+        # parameters, the instance has to be a fully-specified URL for the
+        # instance, not just its name.
+        region_instance_group_managers_delete_instances_request_resource=(
+            compute.RegionInstanceGroupManagersDeleteInstancesRequest(
+                instances=[instance.self_link])))
+  except (google.api_core.exceptions.Forbidden,
+          google.api_core.exceptions.Unauthorized,
+          google.api_core.exceptions.NotFound) as e:
+    print(e)
+    return flask.abort(
+        e.code, f"Error requesting that {mig_name} delete {instance.name}.")
+  except Exception as e:
+    # We'll call any other error here a server error.
+    print(e)
+    return flask.abort(
+        INTERNAL_SERVER_ERROR,
+        f"Error requesting that {mig_name} delete {instance.name}.")
+
+  try:
+    # This is actually an extended operation that you have to poll to get its
+    # status, but we just check the status once because it appears that errors
+    # always show up here and all we just want to return success in marking for
+    # deletion. We don't need to wait for the deletion to actually take place.
+    operation.result()
+  except google.api_core.exceptions.ClientError as e:
+    print(e)
+    # Unpack the actual usable error message
+    msg = (
+        f"Error requesting that {mig_name} delete {instance.name}:"
+        "\n" + "\n".join(
+            [f"{err.code}: {err.message}" for err in e.response.error.errors]))
+    print(msg)
+    # We're not actually totally sure whether this is a client or server error
+    # for the overall request, but let's call it a client error (the only client
+    # here is our VM instances, so I think we can be a bit loose).
+    return flask.abort(BAD_REQUEST, msg)
+
+  success_msg = f"{instance.name} has been marked for deletion by {mig_name}."
+  print(success_msg)
+  return success_msg
+
+
+def should_scale_down(mig_name: str, project: str, region: str):
+  start = time.time()
+  print(f"Polling {mig_name} for stability")
+  while time.time() - start < STABILIZE_TIMEOUT_SECONDS:
+    try:
+      mig = migs_client.get(project=project,
+                            region=region,
+                            instance_group_manager=mig_name)
+    except google.api_core.exceptions.NotFound as e:
+      print(e)
+      return flask.abort(
+          e.code,
+          f"Cannot find {mig_name} in region={region}, project={project}")
+    if mig.status.is_stable:
+      break
+    # We sleep for a random amount of time here to avoid synchronizing callers
+    # waiting for the MIG to be stable.
+    sleep_secs = random.randint(1, 15)
+    print(f"{mig_name} is not stable. Retrying in {sleep_secs} seconds")
+    time.sleep(sleep_secs)
+  else:
+    return flask.abort(GATEWAY_TIMEOUT,
```

Comment:
I guess this should be `REQUEST_TIMEOUT` (408) not `GATEWAY_TIMEOUT` (504)

Reply:
I originally had that, but I investigated further and I think that's not right. `REQUEST_TIMEOUT` is a client error for when the client has opened a connection but failed to send all the data for the request. `GATEWAY_TIMEOUT` is a server error for when: "the server, while acting as a gateway or proxy, did not get a response in time from the upstream server that it needed in order to complete the request". That's pretty much what's happening here (technically it did get a response; the response was just semantically a continuation response).